### PR TITLE
feat(scalasdk): Codegen for Main and AkkaServerlessFactory, #372

### PR DESCRIPTION
**Pull request by patriknw**
First commit is just a refactoring to break out the MainSourceGenerator (in Java sdk)

Refs #372

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -136,7 +136,9 @@ jobs:
       - restore_deps_cache
       - run:
           name: Run tests
-          command: sbt test codegenCore/publishLocal akkaserverless-codegen-scala/publishLocal sdkCore/publishLocal sdkJava/publishLocal sdkScala/publishLocal scripted
+          # FIXME enable scripted again
+          #command: sbt test codegenCore/publishLocal akkaserverless-codegen-scala/publishLocal sdkCore/publishLocal sdkJava/publishLocal sdkScala/publishLocal scripted
```
Comment (patriknw) [inline .circleci/config.yml:140]: I'll have to look into the compilation errors for views first

**Comment by patriknw** [comment]
Ready for final review and merge.

Rebased with latest fixes. Now it's only the AnnotationsProto remaining in the sample.

File: .circleci/config.yml (hunk 1 above)
Comment (patriknw) [reply .circleci/config.yml]: fixed

**Review by octonato — APPROVED** [review]

File: codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/SourceGenerator.scala [hunk 2]
```diff
@@ -57,35 +85,26 @@ object SourceGenerator {
    * user.
    */
   def generateUnmanaged(model: ModelBuilder.Model): Seq[File] = {
-    Seq(File("foo/bar/Baz.scala", "package foo.bar\n\nclass Baz extends AbstractBaz"), generateMain(model)) ++
-    model.services.values
+    val filteredModel = excludedUntilImplemented(model) // FIXME remove
+    Seq(File("foo/bar/Baz.scala", "package foo.bar\n\nclass Baz extends AbstractBaz")) ++
+    MainSourceGenerator.generateUnmanaged(filteredModel) ++
+    filteredModel.services.values
```
Comment (raboof) [inline codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/SourceGenerator.scala:91]: Wouldn't the fact that we use the `filteredModel` here mean we're no longer generating the `ValueEntitySourceGenerator.generateImplementationSkeleton` while we did before?
(my fault for not checking for it in the scripted tests I guess ;) )

File: codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        case service: ModelBuilder.ActionService => s"new ${service.className}(_)"
+        case view: ModelBuilder.ViewService      => s"new ${view.className}(_)"
+      }
+
+    val registrationParameters = entityRegistrationParameters ::: serviceRegistrationParameters
+    s"""|package $mainClassPackageName
+        |
+        |import com.akkaserverless.scalasdk.AkkaServerless
+        |import org.slf4j.LoggerFactory
+        |${componentImports}
```
Comment (raboof) [inline codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala:100]: perhaps add those two extra imports in `generateImports`?

File: codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+        case service: ModelBuilder.ActionService =>
+          List(s".register(${service.providerName}(create${service.className}))")
+
+      }
+      .toList
+      .sorted
+
+    val entityImports = model.entities.values.flatMap { ety =>
+      if (ety.fqn.parent.javaPackage != mainClassPackageName) {
```
Comment (raboof) [inline codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala:148]: this logic is no longer needed here when we use `generateImports` below, I think

File: codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        case view: ModelBuilder.ViewService =>
+          s"create${view.className}: Function[ViewCreationContext, ${view.className}]"
+      }
+
+    val creatorParameters = entityCreators ::: serviceCreators
+    val imports =
+      (List(
+        "com.akkaserverless.scalasdk.AkkaServerless") ++ entityImports ++ serviceImports ++ otherImports ++ contextImports).distinct.sorted
+        .map(pkg => s"import $pkg")
+        .mkString("\n")
```
Comment (raboof) [inline codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala:230]: (could use `generateImports` here)

**Review by raboof — COMMENTED** [review]
I wonder if `excludedUntilImplemented` isn't excluding a bit too much now - otherwise LGTM!

File: codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/SourceGenerator.scala (hunk 2 above)
Comment (patriknw) [reply codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/SourceGenerator.scala]: ok, I moved this filteredModel to MainSourceGenerator instead

File: codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala (hunk 5 above)
Comment (patriknw) [reply codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala]: thanks, done in 94cccf7a5e725516a27f4eb7e150e6949bb8a9f9

File: codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala (hunk 4 above)
Comment (patriknw) [reply codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala]: it wasn't filtering those, only in `typeName`, but I have implemented that in 94cccf7a5e725516a27f4eb7e150e6949bb8a9f9

File: codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala (hunk 3 above)
Comment (patriknw) [reply codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala]: thanks, done in 94cccf7a5e725516a27f4eb7e150e6949bb8a9f9

**Review by raboof — APPROVED** [review]

File: codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala (hunk 4 above)
Comment (raboof) [reply codegen/scala-gen/src/main/scala/com/akkaserverless/codegen/scalasdk/impl/MainSourceGenerator.scala]: ah, nice :+1:

**Pull request merged**